{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e7eceeb8-4592-4785-901c-ba97e839ba07",
   "metadata": {},
   "source": [
    "# Genome Mycobacterium tuberculosis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9f7420d1-a709-4e2a-98da-4da8f2e97bd0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): ...working... done\n",
      "Solving environment: ...working... done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: C:\\Users\\Cristiana\\anaconda3\n",
      "\n",
      "  added / updated specs:\n",
      "    - biopython\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    biopython-1.79             |   py38h294d835_1         2.6 MB  conda-forge\n",
      "    conda-4.11.0               |   py38haa244fe_0        16.9 MB  conda-forge\n",
      "    python_abi-3.8             |           2_cp38           4 KB  conda-forge\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:        19.5 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  biopython          conda-forge/win-64::biopython-1.79-py38h294d835_1\n",
      "  python_abi         conda-forge/win-64::python_abi-3.8-2_cp38\n",
      "\n",
      "The following packages will be SUPERSEDED by a higher-priority channel:\n",
      "\n",
      "  conda              pkgs/main::conda-4.11.0-py38haa95532_0 --> conda-forge::conda-4.11.0-py38haa244fe_0\n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages\n",
      "\n",
      "conda-4.11.0         | 16.9 MB   |            |   0% \n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.conda-4.11.0         | 16.9 MB   |            |   0% \n",
      "conda-4.11.0         | 16.9 MB   | 2          |   2% \n",
      "conda-4.11.0         | 16.9 MB   | 4          |   5% \n",
      "conda-4.11.0         | 16.9 MB   | 7          |   7% \n",
      "conda-4.11.0         | 16.9 MB   | 9          |   9% \n",
      "conda-4.11.0         | 16.9 MB   | #2         |  12% \n",
      "conda-4.11.0         | 16.9 MB   | #4         |  15% \n",
      "conda-4.11.0         | 16.9 MB   | #7         |  18% \n",
      "conda-4.11.0         | 16.9 MB   | ##         |  20% \n",
      "conda-4.11.0         | 16.9 MB   | ##3        |  23% \n",
      "conda-4.11.0         | 16.9 MB   | ##5        |  26% \n",
      "conda-4.11.0         | 16.9 MB   | ##8        |  29% \n",
      "conda-4.11.0         | 16.9 MB   | ###1       |  31% \n",
      "conda-4.11.0         | 16.9 MB   | ###4       |  35% \n",
      "conda-4.11.0         | 16.9 MB   | ###7       |  37% \n",
      "conda-4.11.0         | 16.9 MB   | ####       |  40% \n",
      "conda-4.11.0         | 16.9 MB   | ####2      |  43% \n",
      "conda-4.11.0         | 16.9 MB   | ####5      |  45% \n",
      "conda-4.11.0         | 16.9 MB   | ####7      |  48% \n",
      "conda-4.11.0         | 16.9 MB   | ####9      |  50% \n",
      "conda-4.11.0         | 16.9 MB   | #####2     |  52% \n",
      "conda-4.11.0         | 16.9 MB   | #####4     |  55% \n",
      "conda-4.11.0         | 16.9 MB   | #####6     |  57% \n",
      "conda-4.11.0         | 16.9 MB   | #####9     |  59% \n",
      "conda-4.11.0         | 16.9 MB   | ######1    |  61% \n",
      "conda-4.11.0         | 16.9 MB   | ######3    |  63% \n",
      "conda-4.11.0         | 16.9 MB   | ######5    |  66% \n",
      "conda-4.11.0         | 16.9 MB   | ######8    |  68% \n",
      "conda-4.11.0         | 16.9 MB   | #######    |  70% \n",
      "conda-4.11.0         | 16.9 MB   | #######2   |  73% \n",
      "conda-4.11.0         | 16.9 MB   | #######5   |  75% \n",
      "conda-4.11.0         | 16.9 MB   | #######7   |  78% \n",
      "conda-4.11.0         | 16.9 MB   | ########   |  80% \n",
      "conda-4.11.0         | 16.9 MB   | ########3  |  84% \n",
      "conda-4.11.0         | 16.9 MB   | ########6  |  87% \n",
      "conda-4.11.0         | 16.9 MB   | #########  |  90% \n",
      "conda-4.11.0         | 16.9 MB   | #########3 |  93% \n",
      "conda-4.11.0         | 16.9 MB   | #########6 |  97% \n",
      "conda-4.11.0         | 16.9 MB   | ########## | 100% \n",
      "conda-4.11.0         | 16.9 MB   | ########## | 100% \n",
      "\n",
      "python_abi-3.8       | 4 KB      |            |   0% \n",
      "python_abi-3.8       | 4 KB      | ########## | 100% \n",
      "python_abi-3.8       | 4 KB      | ########## | 100% \n",
      "\n",
      "biopython-1.79       | 2.6 MB    |            |   0% \n",
      "biopython-1.79       | 2.6 MB    |            |   1% \n",
      "biopython-1.79       | 2.6 MB    | ##4        |  24% \n",
      "biopython-1.79       | 2.6 MB    | ####5      |  45% \n",
      "biopython-1.79       | 2.6 MB    | ######5    |  65% \n",
      "biopython-1.79       | 2.6 MB    | ########4  |  84% \n",
      "biopython-1.79       | 2.6 MB    | ########## | 100% \n",
      "Preparing transaction: ...working... done\n",
      "Verifying transaction: ...working... done\n",
      "Executing transaction: ...working... done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda install -c conda-forge biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "87d1309b-d39d-45f7-bc9b-42cdf802fcc6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NCBI\n",
      "\n",
      "ID: AL123456.3\n",
      "Name: AL123456\n",
      "Description: Mycobacterium tuberculosis H37Rv complete genome\n",
      "Database cross-references: BioProject:PRJNA224, BioSample:SAMEA3138326\n",
      "Number of features: 8541\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=BCT\n",
      "/date=27-FEB-2015\n",
      "/accessions=['AL123456']\n",
      "/sequence_version=3\n",
      "/keywords=['complete genome']\n",
      "/source=Mycobacterium tuberculosis H37Rv\n",
      "/organism=Mycobacterium tuberculosis H37Rv\n",
      "/taxonomy=['Bacteria', 'Actinobacteria', 'Corynebacteriales', 'Mycobacteriaceae', 'Mycobacterium', 'Mycobacterium tuberculosis complex']\n",
      "/references=[Reference(title='Deciphering the biology of Mycobacterium tuberculosis from the complete genome sequence', ...), Reference(title='Re-annotation of the genome sequence of Mycobacterium tuberculosis H37Rv', ...), Reference(title='TubercuList--10 years after', ...), Reference(title='Direct Submission', ...), Reference(title='Direct Submission', ...)]\n",
      "/comment=On or before Feb 1, 2013 this sequence version replaced BX842572.1,\n",
      "BX842573.1, BX842574.1, BX842575.1, BX842576.1, BX842577.1,\n",
      "BX842578.1, BX842579.1, BX842580.1, BX842581.1, BX842582.1,\n",
      "BX842583.1, BX842584.1, AL123456.2.\n",
      "On or before Feb 1, 2013 this sequence version replaced\n",
      "gi:41352722, gi:38490165, gi:38490207, gi:41353619, gi:38490250,\n",
      "gi:38684030, gi:38490288, gi:41353667, gi:41353422, gi:41352756,\n",
      "gi:38490319, gi:41352785, gi:38490370, gi:41353971. Note:\n",
      "This annotation is from the TubercuList website, Release 26, Dec\n",
      "2012 (URL: http://tuberculist.epfl.ch) (email:\n",
      "tuberculist@epfl.ch).\n",
      "Seq('TTGACCGATGACCCCGGTTCAGGCTTCACCACAGTGTGGAACGCGGTCGTCTCC...TCG')\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "\n",
    "#Aceder ao NCBI\n",
    "Entrez.email = 'pg45468@alunos.uminho.pt'\n",
    "handle = Entrez.efetch(db = \"nucleotide\", id = \"AL123456.3\", rettype = \"gb\", retmode = \"text\")\n",
    "seq_record = SeqIO.read(handle, \"gb\")\n",
    "\n",
    "#Criar um ficheiro\n",
    "SeqIO.write(seq_record, \"mycobacterium_tuberculosis.gb\", \"genbank\")\n",
    "handle.close()\n",
    "\n",
    "#Aceder aos dados\n",
    "record = SeqIO.read(\"mycobacterium_tuberculosis.gb\", \"genbank\") \n",
    "print (record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e709f3cf-d6d7-475d-87be-0ba5c83f4d20",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
